### PR TITLE
Add `--checkpoint` and `--freeze-encoder` flags in train CLI

### PR DESCRIPTION
**Pull request by shihchengli**
## Description
Addresses #999

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -105,6 +110,12 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
...
     transfer_args = parser.add_argument_group("transfer learning args")
+    transfer_args.add_argument(
+        "--checkpoint",
+        type=Path,
+        nargs="+",
+        help="Location of checkpoint(s) or model file(s) to be loaded for overwriting weights. It can be a path to either a single pretrained model checkpoint (.ckpt) or single pretrained model file (.pt), a directory that contains these files, or a list of path(s) and directory(s). If a directory, will recursively search and predict on all found (.pt) models.",
+    )
     transfer_args.add_argument(
         "--model-frzn",
         help="Path to model checkpoint file to be loaded for overwriting and freezing weights. By default, all MPNN weights are frozen with this option.",
```
Comment (davidegraff) [inline chemprop/cli/train.py:121]: if we have both `--checkpoint` and `--model-frzn` flags, it seems like the logically operate in the same space. That is, both of these flags correspond to a filepath containing model weights to load in, with the only difference that the latter freezes these weights. In this case, I think we should change the `--model-frzn` to be a boolean flag, `--freeze-encoder`, which does exactly that: freezes the loaded encoder weights.

Comment (davidegraff) [reply chemprop/cli/train.py:121]: and FWIW, providing a directory that will be searched for `.pt` and `.ckpt` files seems like a bad idea to me. It seems significantly more idiomatic for a user to directly supply these paths?

Comment (shihchengli) [reply chemprop/cli/train.py:121]: > if we have both `--checkpoint` and `--model-frzn` flags, it seems like the logically operate in the same space. That is, both of these flags correspond to a filepath containing model weights to load in, with the only difference that the latter freezes these weights. In this case, I think we should change the `--model-frzn` to be a boolean flag, `--freeze-encoder`, which does exactly that: freezes the loaded encoder weights.

I also thought about this during the implementation. I think it would be more straightforward to do it the way you suggested. 

> and FWIW, providing a directory that will be searched for `.pt` and `.ckpt` files seems like a bad idea to me. It seems significantly more idiomatic for a user to directly supply these paths?

If a directory is used as input for `--checkpoint`, only the `.pt` files would be found. The reasons for doing providing this capability are (1) if we trained a model on a split with ensemble models, we can just pass the path to the directory to find all those ensemble models instead of passing them explicitly, (2) we can also reuse the `find_models` function we used in `predict.py`, and (3) in v1, we had `--checkpoint_dir` so old users might want this functionality to be implemented in v2 as well.

**Comment by shihchengli** [comment]
I have added the `--freeze-encoder` flag, which can be used with `--checkpoint` to replace `--model-frzn`. The `--model-frzn` is thus depreciated and will be removed in v2.2.

File: chemprop/cli/common.py [hunk 2]
```diff
@@ -191,3 +191,20 @@ def process_common_args(args: Namespace) -> Namespace:
...
+
+    for model_path in model_paths:
+        if model_path.suffix in [".ckpt", ".pt"]:
+            collected_model_paths.append(model_path)
+        elif model_path.is_dir():
+            collected_model_paths.extend(list(model_path.rglob("*.pt")))
+        else:
+            raise ArgumentError(
+                argument=None,
+                message=f"Model path must be a .ckpt, .pt file, or a directory. Got {model_path}",
```
Comment (akshatzalte) [inline chemprop/cli/common.py:207]: message="Expected a .ckpt or .pt file, or a directory. Got {model_path}",

**Review by akshatzalte — COMMENTED** [review]
very minor changes in the help and message statements otherwise LGTM

File: chemprop/cli/train.py [hunk 3]
```diff
@@ -14,7 +14,12 @@
 import torch
 import torch.nn as nn
 
-from chemprop.cli.common import add_common_args, process_common_args, validate_common_args
+from chemprop.cli.common import (
```
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/cli/train.py:17]: that is due to block formatting

File: docs/source/tutorial/cli/train.rst [hunk 4]
```diff
@@ -165,13 +165,11 @@ Advanced Training Methods
...
 
-..  * :code:`--checkpoint-dir <dir>` Directory where the model checkpoint(s) are saved (i.e. :code:`--save_dir` during training of the old model). This will walk the directory, and load all :code:`.pt` files it finds.
-..  * :code:`--checkpoint-path <path>` Path to a model checkpoint file (:code:`.pt` file).
-.. when training the new model. The model architecture of the new model should resemble the architecture of the old model - otherwise some or all parameters might not be loaded correctly. Please note that the old model is only used to initialize the parameters of the new model, but all parameters remain trainable (no frozen layers). Depending on the quality of the old model, the new model might only need a few epochs to train.
+When training the new model, its architecture resemble that of the old model. Depending on the quality of the old model, the new model might only require a few epochs to train.
```
[the remark this answers is not in the document]
Comment (shihchengli) [reply docs/source/tutorial/cli/train.rst:170]: Yes, it is a bit oversimplification.  It also depends on the similarity of the new task, dataset size, and transferability of learned representation. How about change it into
```suggestion
When training the new model, its architecture **must** resemble that of the old model. Depending on the similarity of the tasks and datasets, as well as the quality of the old model, the new model might require fewer epochs to achieve optimal performance compared to training from scratch.
```

File: chemprop/cli/common.py [hunk 5]
```diff
@@ -191,3 +191,20 @@ def process_common_args(args: Namespace) -> Namespace:
 
 def validate_common_args(args):
     pass
+
+
+def find_models(model_paths: list[Path]):
```
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/cli/common.py:196]: added

**Review by akshatzalte — APPROVED** [review]

File: chemprop/cli/train.py [hunk 6]
```diff
@@ -842,7 +864,34 @@ def train_model(
...
+                model.apply(
+                    lambda m: setattr(m, "p", args.dropout)
+                    if isinstance(m, torch.nn.Dropout)
+                    else None
+                )
+
+            # TODO: model_frzn is deprecated and then remove in v2.2
+            if args.model_frzn or args.freeze_encoder:
+                model.message_passing.apply(lambda module: module.requires_grad_(False))
+                model.message_passing.apply(
```
Comment (KnathanM) [inline chemprop/cli/train.py:886]: #952 also touches this code in build_model, but it shouldn't be too hard to update that PR to fit this one.

File: chemprop/cli/train.py [hunk 7]
```diff
@@ -842,7 +864,34 @@ def train_model(
...
+            )
+            model_path = model_paths[model_idx] if args.checkpoint else args.model_frzn
+            model = mpnn_cls.load_from_file(model_path)
+
+            if args.checkpoint:
+                model.apply(
+                    lambda m: setattr(m, "p", args.dropout)
+                    if isinstance(m, torch.nn.Dropout)
+                    else None
+                )
```
Comment (KnathanM) [inline chemprop/cli/train.py:881]: Why don't we use the dropout value that is in the loaded model file?

File: chemprop/cli/train.py [hunk 8]
```diff
@@ -110,6 +115,15 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
...
 
     transfer_args = parser.add_argument_group("transfer learning args")
+    transfer_args.add_argument(
+        "--checkpoint",
+        type=Path,
+        nargs="+",
+        help="Path to checkpoint(s) or model file(s) for loading and overwriting weights. Accepts a single pre-trained model checkpoint (.ckpt), a single model file (.pt), a directory containing such files, or a list of paths and directories. If a directory is provided, it will recursively search for and use all (.pt) files found for prediction.",
+    )
+    transfer_args.add_argument(
+        "--freeze-encoder", action="store_true", help="Whether to freeze the MPNN weights."
```
Comment (KnathanM) [inline chemprop/cli/train.py:125]:
```suggestion
        "--freeze-encoder", action="store_true", help="Freeze the message passing layer from the checkpoint model (specified by ``--checkpoint``."
```

File: chemprop/cli/train.py [hunk 9]
```diff
@@ -118,7 +132,7 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
         "--frzn-ffn-layers",
         type=int,
         default=0,
-        help="Freeze the first ``n`` layers of the FFN from the checkpoint model (specified by ``model-frzn``).",
+        help="Freeze the first ``n`` layers of the FFN from the checkpoint model (specified by ``checkpoint`` with `freeze-encoder`).",
```
Comment (KnathanM) [inline chemprop/cli/train.py:135]:
```suggestion
        help="Freeze the first ``n`` layers of the FFN from the checkpoint model (specified by `--`checkpoint``). The message passing layer should also be frozen with ``--freeze-encoder``).",
```

Comment (KnathanM) [inline chemprop/cli/train.py:135]: Should we add a check to require that `--checkpoint` is not None if `--freeze-encoder` is specified? Should we add a check to require that `--checkpoint` and `--freeze-encoder` are not None if `--frzn-ffn-layers` is specified?

File: chemprop/cli/train.py [hunk 10]
```diff
@@ -110,6 +115,15 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
     # )
 
     transfer_args = parser.add_argument_group("transfer learning args")
+    transfer_args.add_argument(
+        "--checkpoint",
```
Comment (KnathanM) [inline chemprop/cli/train.py:119]: Will we eventually add the option to restart training from a .ckpt file? Right now the models specified here are loaded using `load_from_file` so the training information (gradients, etc.) are not loaded. So this argument isn't for restarting training from a checkpoint. Maybe later we will add another argument like resume_training` that will require all the paths in `args.checkpoint` to be .ckpt files and will use `load_from_checkpoint` instead.

File: chemprop/cli/train.py (hunk 7 above)
Comment (shihchengli) [reply chemprop/cli/train.py:881]: For model fine-tuning, I think it makes sense to adjust the dropout rate to suit the new task or dataset, as the previous rate was optimized for the pre-trained dataset and may not be suitable for the new dataset.

File: chemprop/cli/train.py (hunk 10 above)
Comment (shihchengli) [reply chemprop/cli/train.py:119]: Yes, that is the plan. We left the addition of the `--resume-experiment` flag for v2.1. We can consider assigning someone to work on it.

File: chemprop/cli/train.py [hunk 11]
```diff
@@ -493,6 +493,24 @@ def process_train_args(args: Namespace) -> Namespace:
...
+            message="`--freeze-encoder` can only be used when `--checkpoint` is used.",
+        )
+
+    if args.frzn_ffn_layers > 0:
+        if args.checkpoint is None and args.model_frzn is None:
+            raise ArgumentError(
+                argument=None,
+                message="`--frzn-ffn-layers` can only be used when `--checkpoint` or `--model-frzn` (depreciated in v2.1) is used.",
+            )
+        if args.checkpoint is not None and not args.freeze_encoder:
```
Comment (KnathanM) [inline chemprop/cli/train.py:512]:
```suggestion
        if not args.freeze_encoder:
```
Do you need to check if the checkpoint is not None?

Comment (shihchengli) [reply chemprop/cli/train.py:512]: The `--model-frzn` flag is still keeped in v2.1. When using this flag, it is not necessary to use `--freeze_encoder` to freeze the MPNN. Once the `--model-frzn` flag is removed, we can then use the code you suggested here.

**Review by KnathanM — APPROVED** [review]
